packages/itwinui-react/src/core/Menu/MenuItem.tsx

```diff
@@ -139,112 +190,143 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
+            }
+          },
+        )}
+        role={role}
+        tabIndex={disabled || role === 'presentation' ? undefined : -1}
+        aria-selected={isSelected}
+        aria-haspopup={subMenuItems.length > 0 ? 'true' : undefined}
+        aria-controls={subMenuItems.length > 0 ? submenuId : undefined}
+        aria-expanded={subMenuItems.length > 0 ? popover.open : undefined}
+        aria-disabled={disabled}
+        {...(subMenuItems.length === 0
+          ? { ...handlers, ...rest }
+          : popover.getReferenceProps({ ...handlers, ...rest }))}
+      >
+        {startIcon && (
+          <ListItem.Icon as='span' aria-hidden>
+            {startIcon}
+          </ListItem.Icon>
+        )}
+        <ListItem.Content>
+          <div>{children}</div>
+          {sublabel && <ListItem.Description>{sublabel}</ListItem.Description>}
+        </ListItem.Content>
+        {!endIcon && subMenuItems.length > 0 && (
+          <ListItem.Icon as='span' aria-hidden>
+            <SvgCaretRightSmall />
+          </ListItem.Icon>
+        )}
+        {endIcon && (
+          <ListItem.Icon as='span' aria-hidden>
+            {endIcon}
+          </ListItem.Icon>
+        )}
+      </ListItem>
 
       {subMenuItems.length > 0 && popover.open && (
-        <Portal>
-          <MenuItemContext.Provider
-            value={{ ref: menuItemRef, setIsNestedSubmenuVisible }}
-          >
-            <Menu
-              setFocus={focusOnSubmenu}
-              ref={popover.refs.setFloating}
-              {...popover.getFloatingProps({
-                id: submenuId,
-                onPointerMove: () => {
-                  // pointer might move into a nested submenu and set isSubmenuVisible to false,
-                  // so we need to flip it back to true when pointer re-enters this submenu.
-                  setIsSubmenuVisible(true);
-                },
-              })}
+        <FloatingNode id={nodeId}>
+          <Portal>
+            <MenuItemContext.Provider
+              value={{
+                setActiveIndex,
+                listItemsRef,
+                setHasFocusedNodeInSubmenu,
+              }}
             >
-              {subMenuItems}
-            </Menu>
-          </MenuItemContext.Provider>
-        </Portal>
+              <Menu
+                setFocus={false}
+                ref={popover.refs.setFloating}
+                {...popover.getFloatingProps({
+                  id: submenuId,
+                  onPointerMove: () => {
+                    // pointer might move into a nested submenu and set isSubmenuVisible to false,
+                    // so we need to flip it back to true when pointer re-enters this submenu.
+                    setIsSubmenuVisible(true);
+                  },
+                })}
+              >
+                {subMenuItems.map((item, index) =>
+                  React.cloneElement(item, {
+                    key: index,
+                    ['data-menu-child-index']: index,
```

Comment:
Two comments here:
1. Data attributes should be namespaced. Something like `data-iui-index` would suffice.
2. While it works for `MenuItem`, cloning elements like this is not super reliable. It would be better to use something like context. We could maybe save this improvement for when we deprecate `MenuItem` in favor of `DropdownMenu.Item` _([related conversation](https://github.com/iTwin/iTwinUI/pull/1942#discussion_r1561066189))_.

Reply:
For now, I scoped the data attribute to `iui` by renaming it to `data-iui-index` ([`c3f0946`](https://github.com/iTwin/iTwinUI/pull/1942/commits/c3f09463b8bc7313883a36527d56506080cf2adc) + https://github.com/iTwin/iTwinUI/pull/1942/commits/8016f5bfcd8961c67ddb6a968094324157cd48e6).

I added an after PR TODO to try replacing the `cloneElement` with contexts.

Reply:
`cloneElement` and data attributes were replaced with context in [`142a531`](https://github.com/iTwin/iTwinUI/pull/1942/commits/142a531fa9b3dc358340aec0485405d217557419) in order to avoid reading from refs (and DOM elements) during render.

The other part of my comment (deprecating `MenuItem`) can be kept as an after-PR TODO.